csrc/kernel_cache.cpp

```diff
@@ -746,20 +745,91 @@ std::vector<at::Tensor> FusionKernelRuntime::runWithInputs(
+      const std::vector<SegmentedGroup*>& group_run_order,
+      std::unordered_map<Val*, const ArgAbstract*>& tensor_map)
+      : fusion_args_(args),
+        group_run_order_(group_run_order),
+        tensor_map_(tensor_map) {
+    setLastUsedSegmentID();
+  }
+  template <typename T>
+  void run(
+      const std::vector<Val*>& group_outputs,
+      const T& group_runtime_outputs,
+      const int group_id) {
+    addOutputsToArgsAndTensorMap(group_outputs, group_runtime_outputs);
+    deleteUnusedArgs(group_id);
+  }
 
-  auto update_outputs = [&args, &tensor_map](
-                            auto group_outputs, auto group_runtime_outputs) {
+ private:
+  KernelArgumentHolder& fusion_args_;
+  const std::vector<SegmentedGroup*>& group_run_order_;
+  // map from val to args
+  std::unordered_map<Val*, const ArgAbstract*>& tensor_map_;
+  // map segment_id to vector of fusion vals lastly used at this segment
+  std::unordered_map<int, std::vector<Val*>> vals_last_used_at_segment_;
+
+  void setLastUsedSegmentID() {
+    // map val to segment_id where arg is lastly used
+    std::unordered_map<Val*, int> last_used_segment_map;
+    const int n_groups = group_run_order_.size();
+    // only need to set lifetime of vals if there are more than 3 groups
+    if (n_groups >= 3) {
+      // start from the 2nd group, since the input of the first group is always
+      // the global input and its outputs are always used by at least one of the
+      // following groups
+      for (int group_id = 1; group_id < n_groups; ++group_id) {
+        auto group_to_run = group_run_order_.at(group_id);
+        // set/update life of vals in inputs of this group
+        for (auto val : group_to_run->inputs()) {
+          // skip fusion inputs since if a Val is a fusion input, it means it's
+          // a user input.
+          if (!val->isFusionInput()) {
+            last_used_segment_map[val] = group_id;
+          }
+        }
+        // set/update life of vals in outputs of this group
+        // skip the last group since its outputs are always the global outputs
+        if (group_id < n_groups - 1) {
+          for (auto val : group_to_run->outputs()) {
+            // skip fusion outputs, outputs may be used by other fusions or code
+            if (!val->isFusionOutput()) {
+              last_used_segment_map[val] = group_id;
+            }
+          }
+        }
+      }
+      // convert to vals_last_used_at_segment_, so we don't need to iterate over
+      // all vals when erasing
+      for (auto item : last_used_segment_map) {
+        vals_last_used_at_segment_[item.second].push_back(item.first);
+      }
+    }
+  }
+  void deleteUnusedArgs(int group_id) {
+    // erase args corresponding to vals lastly used in this segment
+    if (group_id >= 1 && vals_last_used_at_segment_.count(group_id)) {
+      for (auto val : vals_last_used_at_segment_[group_id]) {
+        fusion_args_.erase(tensor_map_.at(val)->copy_unique_ptr());
```

Comment:
This interface seems strange to me. Why do we need to make a copy of a unique ptr to delete something?

Reply:
I changed the interface.